cmd/cosign/cli/verify/verify_bundle.go

```diff
@@ -0,0 +1,163 @@
+		}
+	}
+
+	trustedmaterial := &verifyTrustedMaterial{TrustedMaterial: trustedroot}
+
+	// See if we need to wrap trusted root with provided key
+	if keyRef != "" {
+		signatureVerifier, err := sigs.PublicKeyFromKeyRef(ctx, keyRef)
+		if err != nil {
+			return err
+		}
+
+		newExpiringKey := root.NewExpiringKey(signatureVerifier, time.Time{}, time.Time{})
+		trustedmaterial.keyTrustedMaterial = root.NewTrustedPublicKeyMaterial(func(_ string) (root.TimeConstrainedVerifier, error) {
+			return newExpiringKey, nil
+		})
+	} else if sk {
+		s, err := pivkey.GetKeyWithSlot(slot)
+		if err != nil {
+			return fmt.Errorf("opening piv token: %w", err)
+		}
+		defer s.Close()
+		signatureVerifier, err := s.Verifier()
+		if err != nil {
+			return fmt.Errorf("loading public key from token: %w", err)
+		}
+
+		newExpiringKey := root.NewExpiringKey(signatureVerifier, time.Time{}, time.Time{})
+		trustedmaterial.keyTrustedMaterial = root.NewTrustedPublicKeyMaterial(func(_ string) (root.TimeConstrainedVerifier, error) {
+			return newExpiringKey, nil
+		})
+	}
+
+	identityPolicies := []verify.PolicyOption{}
+
+	verificationMaterial := bundle.GetVerificationMaterial()
+
+	if verificationMaterial == nil {
+		return fmt.Errorf("no verification material in bundle")
+	}
+
+	if verificationMaterial.GetPublicKey() != nil {
+		identityPolicies = append(identityPolicies, verify.WithKey())
+	} else {
+		sanMatcher, err := verify.NewSANMatcher(certIdentity, certIdentityRegexp)
+		if err != nil {
+			return err
+		}
+
+		extensions := certificate.Extensions{
+			Issuer:                   certOIDCIssuer,
+			GithubWorkflowTrigger:    githubWorkflowTrigger,
+			GithubWorkflowSHA:        githubWorkflowSHA,
+			GithubWorkflowName:       githubWorkflowName,
+			GithubWorkflowRepository: githubWorkflowRepository,
+			GithubWorkflowRef:        githubWorkflowRef,
+		}
+
+		certIdentity, err := verify.NewCertificateIdentity(sanMatcher, extensions)
+		if err != nil {
+			return err
+		}
+
+		identityPolicies = append(identityPolicies, verify.WithCertificateIdentity(certIdentity))
+	}
+
+	// Make some educated guesses about verification policy
+	verifierConfig := []verify.VerifierOption{}
+
+	if len(trustedroot.RekorLogs()) > 0 && !ignoreTlog {
+		verifierConfig = append(verifierConfig, verify.WithTransparencyLog(1), verify.WithObserverTimestamps(1))
+	}
+
+	if len(trustedroot.TimestampingAuthorities()) > 0 && useSignedTimestamps {
+		verifierConfig = append(verifierConfig, verify.WithSignedTimestamps(1))
+	}
+
+	if !ignoreSCT {
+		verifierConfig = append(verifierConfig, verify.WithSignedCertificateTimestamps(1))
+	}
+
+	if ignoreTlog && !useSignedTimestamps {
+		verifierConfig = append(verifierConfig, verify.WithoutAnyObserverTimestampsUnsafe())
+	}
+
+	// Perform verification
+	payload, err := payloadBytes(artifactRef)
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewBuffer(payload)
+
+	sev, err := verify.NewSignedEntityVerifier(trustedmaterial, verifierConfig...)
```

Comment:
Originally, what I had in mind was decomposing the bundle into the `CheckOpts` and OCI structs so that we could call [VerifyBlobSignature](https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/verify/verify_blob.go#L293). The concern was a lack of parity between the two verification policies.

Given that this is guarded behind a flag, it wouldn't be a breaking change to have a different verification path. And maybe I'm overthinking this, that there isn't actually a significant gap between the two verification policies - keys was one concern, but it seems like you've fully supported that here.

WDYT, between either using sigstore-go's verification library or continuing to use Cosign's with bundle decomposition?

Reply:
That was the path I originally started going down - decomposing the bundle and using cosign's existing `VerifyBlobSignature()` - similar to the technique we used to add signing support. But it's a fair amount of work to decompose both the bundle and the trusted root into the format cosign is expecting, which would look very similar to code that's already in sigstore-go. Since I think we want to eventually get sigstore-go into cosign (and have more parts of cosign use sigstore-go in the future) I jumped ahead a bit in this pull request.

Reply:
I agree, I'm happy to see us cut over now. I've been looking over `VerifyBlobSignature` to see if there are any differences that we'd be missing, but I don't see anything concerning.

Online verification for Rekor is missing, but that's a good thing, I want to remove this. Bundles always have inclusion proofs and online lookups by default isn't a great design.

There is no support for verifying with certificate chains bundled in the verification material, but I know that's blocked on support in sigstore-go, and it's simple enough to work around with a trust root file.